### PR TITLE
initial commit: recommendations bh

### PR DESCRIPTION
**Pull request by maxreid-openai**
Initial commit for recommendations BH 12/12

### CODE REVIEW

File: 11-recommendation/helper_functions.py [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+
+    return selected_records
+
+
+###  Page 1: 
+
+## Expand Query
+# Structured Outputs to limit the LLM's response to specific categoriess
+
+#
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:205]: nit: extra comment

Comment (ericning-o) [inline 11-recommendation/helper_functions.py:205]: nit: extra comment

File: 11-recommendation/helper_functions.py [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+
+            6. **Output Format**:
+            - Present your findings in a JSON format as shown in the example below, which is delimited by `</example>` tags.
+            - Ensure the JSON is correctly structured and valid.
+
+            7. **Additional Behavioral Notes**:
+            - **Relevance**: Ensure all categories and descriptions are tailored to the user's input and past purchases.
+            - **Tone**: Maintain a fun and playful tone to engage the user.
+            - **Quantity**: Do not exceed 7 categories in your final list.
+
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:301]: nit remove this extra whitespace

File: 11-recommendation/helper_functions.py [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+    CLEANING_PRESSURE_WASHERS = "Cleaning: Pressure Washers"
+    OUTDOOR_GRILLS_OUTDOOR_COOKING = "Outdoor: Grills and Outdoor Cooking"
+    OUTDOOR_LAWN_MOWERS = "Outdoor: Lawn Mowers"
+    OUTDOOR_PATIO_FURNITURE = "Outdoor: Patio Furniture"
+    OUTDOOR_FIRE_PITS = "Outdoor: Fire Pits"
+    SAFETY_PERSONAL_PROTECTIVE_EQUIPMENT = "Safety: Personal Protective Equipment"
+    SAFETY_FIRST_AID_SUPPLIES = "Safety: First Aid Supplies"
+    SAFETY_FIRE_EXTINGUISHERS = "Safety: Fire Extinguishers"
+    SAFETY_SAFETY_SIGNS = "Safety: Safety Signs"
+class Category(BaseModel):
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:247]: nit: add line above

File: 11-recommendation/helper_functions.py [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+            - Only include categories that are directly relevant to the user's input.
+            - Bias towards categories that are complementary to the user's past purchases. However, if the user's goal is not related to their past purchases, do not include any categories related to their past purchases.
+
+            3. **Craft Engaging Descriptions**:
+            - For each selected category, write a fun and playful description to excite the user about their project.
+            - ONLY if applicable, mention how their past purchases contribute to their progress, highlighting that they're already on their way.
+
+            
+            **Best Practices**:
+            4. **Include Essential and Complementary Categories**:
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:287]: Not sure you need to number them anymore for best practices. You should just have it in a list for 56? Which parts are best practices?

File: 11-recommendation/helper_functions.py [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+            - Incorporate both categories that are essential for the user's project and those that offer complementary products (e.g., suggesting outdoor furniture for someone building a deck).
+
+            5. **Reference Past Purchases**:
+            - Utilize the past purchases provided, which are delimited by `<past_purchase>` tags, to personalize recommendations and reinforce progress. If the past purchase is not relevant, do not include it in the description.
+
+            6. **Output Format**:
+            - Present your findings in a JSON format as shown in the example below, which is delimited by `</example>` tags.
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:294]: Interesting, do you actually need to state this if you have structured outputs set?

File: 11-recommendation/helper_functions.py [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+                {"role": "system", "content": system_prompt},
+                {"role": "user", "content": query}
+            ],
+            response_format=CategoryRecommendations
+
+        )
+        # Extract the generated text
+        expanded_query = response.choices[0].message.content
+        logger.info("Query expanded successfully using LLM.")
+        print(expanded_query)
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:343]: rm print?

File: 11-recommendation/helper_functions.py [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+        logger.info(f"Querying Qdrant with category: '{category}' with query: '{query}'")
+        # Start of Selection
+        if not query:
+            logger.error("Empty query string. Cannot perform query.")
+            return []
+        else:   
+            # Generate embeddings for the query
+            query_vector = generate_embeddings(query)
+
+
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:383]: nit: rm extra line

File: 11-recommendation/helper_functions.py [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+        {
+            "id": "f1e4b7ec-1159-4600-a17e-e817f6fa6c83",
+            "product_name": "RuggedGrip Self-Locking Hook",
+            "category": "Hardware: Locks and Latches",
+            "price": 40.84,
+            "predicted_score": 0.83
+        }
+        ]
+
+
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:490]: nit: rm extra line

File: 11-recommendation/helper_functions.py [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+            "predicted_score": 1.0
+        },
+        {
+            "id": "f1e4b7ec-1159-4600-a17e-e817f6fa6c83",
+            "product_name": "RuggedGrip Self-Locking Hook",
+            "category": "Hardware: Locks and Latches",
+            "price": 40.84,
+            "predicted_score": 0.83
+        }
+        ]
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:488]: nit: weird spacing, shouldn't the bracket be one indented in?

File: 11-recommendation/helper_functions.py [hunk 10]
```diff
@@ -0,0 +1,530 @@
...
+    - Get the user excited about the recommended product.
+    - Reflect their preferences implied by their purchase history. You can reference categories in the past purchases.
+    - Respond in exactly two sentences.
+    - Be concise, positive, and engaging.
+    """
+
+    response = openai_client.chat.completions.create(
+        model="gpt-4o",
+        messages=[{"role":"system","content":prompt}],
+        max_tokens=150,
```
Comment (ericning-o) [inline 11-recommendation/helper_functions.py:525]: Do we need the max_tokens and temp set here? Is it useful?

File: 11-recommendation/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,59 @@
+altair==5.5.0
```
Comment (ericning-o) [inline 11-recommendation/requirements.txt:1]: not that it matters too much, but do we actually need every lib here? like `GitPython` `h11` etc

File: 11-recommendation/upload-to-qdrant.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    data_entries = []
+    try:
+        with open(csv_file_path, mode='r', encoding='utf-8') as csvfile:
+            reader = csv.DictReader(csvfile)
+            for row in reader:
+                # Convert data types as necessary
+                row['price'] = float(row['price']) if row['price'] else None
+                 # Parse the embedding vector from string to list of floats
+                embedding_str = row.get('embedded_description')  # Replace 'embedding' with your actual column name
```
Comment (ericning-o) [inline 11-recommendation/upload-to-qdrant.py:30]: nit: weird comment? can remove

File: 11-recommendation/upload-to-qdrant.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    Reads data from a CSV file and returns a list of dictionaries.
+    """
+    data_entries = []
+    try:
+        with open(csv_file_path, mode='r', encoding='utf-8') as csvfile:
+            reader = csv.DictReader(csvfile)
+            for row in reader:
+                # Convert data types as necessary
+                row['price'] = float(row['price']) if row['price'] else None
+                 # Parse the embedding vector from string to list of floats
```
Comment (ericning-o) [inline 11-recommendation/upload-to-qdrant.py:29]: nit: this comment has weird spacing

File: 11-recommendation/upload-to-qdrant.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+            else:
+                logger.warning(f"Missing embedding for row {idx}. Converting to empty list.")
+                embedded_description = []
+        entry['embedded_description'] = embedded_description
+
+        point = PointStruct(
+            id=idx,
+            vector=embedded_description,
+            payload={
+                    # Start of Selection
```
Comment (ericning-o) [inline 11-recommendation/upload-to-qdrant.py:90]: don't need this comment

File: .gitignore [hunk 15]
```diff
@@ -4,4 +4,8 @@ __pycache__/
 __pycache__/
 
 3-4o_mini_fine_tuning/evals/*
-3-4o_mini_fine_tuning/data/*
\ No newline at end of file
+3-4o_mini_fine_tuning/data/*
+
+11-recommendation/images
+
+.venv/
```
Comment (ericning-o) [inline .gitignore:11]: Are we actually using venv here? Should it be in the instructions readme?

File: 11-recommendation/pages/page_2_explainable_recommendations.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+        st.session_state.explanations = {}
+
+    # Display Previous Purchases
+    st.subheader("Previous Purchases")
+    past_purchases = get_past_purchases()
+    with st.expander("Show Previous Purchases"):
+        for purchase in past_purchases:
+            st.write(f"{purchase['product_name']} - Purchased {purchase['quantity']} times")
+
+    # Button to Generate Recommendations
```
Comment (ericning-o) [inline 11-recommendation/pages/page_2_explainable_recommendations.py:29]: what button 🤔

File: 11-recommendation/pages/page_1_semantic_search.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+    # Initialize session state
+    if 'expanded_query' not in st.session_state:
+        st.session_state.expanded_query = None
+
+    if 'search_results' not in st.session_state:
+        st.session_state.search_results = {}
+
+    # Load distinct categories and past purchases
+
```
Comment (ericning-o) [inline 11-recommendation/pages/page_1_semantic_search.py:28]: nit: rm unnecessary whitespace

**Review by ericning-o — APPROVED** [review]
generally looks good, just a bunch of nits and cleanup you could do

File: .gitignore (hunk 15 above)
Comment (maxreid-openai) [reply .gitignore:11]: good call!
